core/unix/signal_private.h

```diff
@@ -392,6 +392,14 @@ typedef struct _sighand_info_t {
      */
     kernel_sigaction_t *action[SIGARRAY_SIZE];
     bool we_intercept[SIGARRAY_SIZE];
+    /* For handling masked-for-app-but-not-for-DR signals.  Any time we receive
+     * a signal in a thread for which it is blocked, we need to know whether it was
+     * a "process"-wide signal and whether some other thread has it unblocked.
+     * To avoid heavyweight locks every time, we keep an atomic-access counter of
+     * unmasked threads for each signal, for the CLONE_SIGHAND group (typically
+     * whole process).
+     */
+    int threads_unmasked[SIGARRAY_SIZE];
```

Comment:
Yes, it is potentially fragile.  The assert at exit helped and with the logging I added I was able to debug several issues without too much trouble, but it will always be a little messy.  But, I couldn't come up with a cleaner solution so I think this is the best we have for now.

Added a comment by app_sigblocked.